src/SessionSetup/RetrieveInstancesWidget.cpp

```diff
@@ -171,4 +173,25 @@ void RetrieveInstancesWidget::OnInstancesLoadingReturned(
   emit LoadingSuccessful(loading_result.value());
 }
 
+void RetrieveInstancesWidget::OnReloadButtonClicked() {
+  // Initial loading failed
+  if (ui_->projectComboBox->count() == 0) {
+    InitialLoad(std::nullopt);
+    return;
+  }
+
+  std::optional<Project> selected_project;
+  if (ui_->projectComboBox->currentData().canConvert<Project>()) {
+    selected_project = ui_->projectComboBox->currentData().value<Project>();
+  }
+
+  emit LoadingStarted();
+  retrieve_instances_->LoadInstancesWithoutCache(selected_project, GetInstanceListScope())
+      .Then(main_thread_executor_.get(), [widget = QPointer<RetrieveInstancesWidget>(this)](
+                                             const ErrorMessageOr<QVector<Instance>>& load_result) {
+        if (widget == nullptr) return;
+        widget->OnInstancesLoadingReturned(load_result);
+      });
```

Comment:
Is the main_thread_executor_ local to this object? If yes then you can skip the lifetime check because it's enforced implicitly.

Reply:
I dug into Executer and indeed, you built in the lifetime checks. Thanks beckerhe from the past. I did change it and added a comment

Reply:
I still have to get rid of the shared_ptr which makes that highly confusing.